src/nr_openai_observability/langchain_callback.py

```diff
@@ -0,0 +1,212 @@
+        tags = {error: str(error)}
+        span = self.spans_stack.pop()
+        assert span["attributes"]["name"] == "LlmChain"
+        self.finish_and_record_span(span, tags)
+
+    def on_tool_start(
+        self, serialized: Dict[str, Any], input_str: str, **kwargs: Any
+    ) -> Any:
+        """Run when tool starts running."""
+        tags = {
+            "tool_name": serialized.get("name"),
+            "tool_description": serialized.get("description"),
+            "tool_input": input_str,
+        }
+        self.spans_stack.append(self.create_span(name="LlmTool", tags=tags))
+
+    def on_tool_end(self, output: str, **kwargs: Any) -> Any:
+        """Run when tool ends running."""
+        tags = {
+            "tool_output": output,
+            "tool_invocation_counter": self.get_and_update_tool_invocation_counter(),
+        }
+        span = self.spans_stack.pop()
+
+        assert span["attributes"]["name"] == "LlmTool"
+        tool_name = kwargs.get("name")
+        assert span["attributes"]["tool_name"] == tool_name
+
+        self.finish_and_record_span(span, tags)
+
+    def on_tool_error(
+        self, error: Union[Exception, KeyboardInterrupt], **kwargs: Any
+    ) -> Any:
+        """Run when tool errors."""
+        tags = {
+            "error": error,
+        }
+        span = self.spans_stack.pop()
+        assert span["attributes"]["name"] == "LlmTool"
+        tool_name = kwargs.get("name")
+        assert span["attributes"]["tool_name"] == tool_name
+
+        self.finish_and_record_span(span, tags)
+
+    def on_text(self, text: str, **kwargs: Any) -> Any:
+        """Run on arbitrary text."""
+
+    def on_agent_action(self, action: AgentAction, **kwargs: Any) -> Any:
+        """Run on agent action."""
+
+    def on_agent_finish(self, finish: AgentFinish, **kwargs: Any) -> Any:
+        """Run on agent end."""
+
+    def parent_id_callback(self) -> str:
+        parent_span = self.spans_stack[-1] if self.spans_stack else None
+        return parent_span["id"] if parent_span else None
+
+    def create_span(
+        self,
+        name: Optional[str] = None,
+        tags: Optional[Dict[str, Any]] = None,
+        guid: Optional[str] = None,
+        trace_id: Optional[str] = None,
+        parent_id: Optional[str] = None,
+        start_time_ms: Optional[int] = None,
+        duration_ms: Optional[int] = None,
+    ):
+        if parent_id is None:
+            parent_span = self.spans_stack[-1] if self.spans_stack else None
+            parent_id = parent_span["id"] if parent_span else None
+
+        if self.langchain_callback_metadata:
+            tags = tags or {}
+            tags.update(self.langchain_callback_metadata)
+
+        if not trace_id and "newrelic" in sys.modules:
+            import newrelic.agent # type: ignore
+
+            trace_id = getattr(newrelic.agent.current_transaction(), "trace_id", None)
+
+        trace_id = trace_id or self.trace_id
```

Comment:
If a user created one callback handler object and attempted to reuse it, it seems like they would get the same `trace_id` as their previous request (if they weren't also using the agent and relied on `self.trace_id`).

Reply:
Thats a good callout, except if instructing the users to recreate the object to every request, any idea how can i be solved?
From the callback perspective, the highest abstraction is a chain but we may have multiple chains per transaction so we can be depend on the on_chain_start to reset the trace_id.